libs/payments/authenticated_payment_state.go

```diff
@@ -0,0 +1,82 @@
+package payments
+
+import (
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/shopspring/decimal"
+	"golang.org/x/exp/slices"
+)
+
+type PaymentDetails struct {
+	Amount    decimal.Decimal `json:"amount" valid:"required"`
+	To        string          `json:"to" valid:"required"`
+	From      string          `json:"from" valid:"required"`
+	Custodian string          `json:"custodian" valid:"in(uphold|gemini|bitflyer)"`
+	PayoutID  string          `json:"payoutId" valid:"required"`
+	Currency  string          `json:"currency" valid:"required"`
+}
+
+// PaymentAuthorization represents a single authorization from a payment authorizer indicating that
+// the payout represented by a document ID should be processed
+type PaymentAuthorization struct {
+	KeyID      string `json:"keyId" valid:"required"`
+	DocumentID string `json:"documentId" valid:"required"`
+}
+
+type AuthenticatedPaymentState struct {
+	PaymentDetails
+	Status         PaymentStatus
+	Authorizations []PaymentAuthorization `json:"authorizations"`
+	DryRun         *string                `json:"dryRun"`
+	LastError      *PaymentError
+	DocumentID     string
+}
+
+func (p *AuthenticatedPaymentState) GenerateIdempotencyKey(namespace uuid.UUID) uuid.UUID {
+	return uuid.NewSHA1(
+		namespace,
+		[]byte(fmt.Sprintf(
+			"%s%s%s%s%s%s",
+			p.To,
+			p.From,
+			p.Currency,
+			p.Amount,
+			p.Custodian,
+			p.PayoutID,
+		)),
+	)
+}
+
+func (t *AuthenticatedPaymentState) NextStateValid(nextState PaymentStatus) bool {
+	if t.Status == nextState {
+		return true
+	}
+	// New transaction state should be present in the list of valid next states for the current
+	// state.
+	if !slices.Contains(t.Status.GetValidTransitions(), nextState) {
```

Comment:
suggestion (non-blocking): this could written as `return slices.Contains(t.Status.GetValidTransitions(), nextState)` to be more concise